File: BE/src/model/course.model.ts
import mongoose, { Document, Schema } from "mongoose";

export type CourseStatus = "not_yet" | "in_progress" | "complete";

export type JLPTLevel = "N5" | "N4" | "N3" | "N2" | "N1";

export interface ICourse extends Document {
  name: string;
  description?: string;
  status: CourseStatus;
  startDate: Date;
  endDate: Date;
  createdBy: string;
  homeroomTeacherId: Schema.Types.ObjectId;
  homeroomTeacher: string;
  session: number;
  capacity: number;
  enrolledCount: number;
  members: {
    userId: Schema.Types.ObjectId;
    role: "student" | "teacher";
    enrolledAt: Date;
    deletedAt?: Date | null;
    deletedBy?: string | null;
  }[];
}

const courseSchema = new Schema<ICourse>(
  {
    name: { type: String, required: true },
    description: String,

    status: {
      type: String,
      enum: ["not_yet", " in_progress", "complete"],
      default: "not_yet",
      required: true,
    },

    startDate: {type: Date, required: true},
    endDate: {type: Date, required: true},
    createdBy: { type: String, required: true },
    homeroomTeacherId: { type: Schema.Types.ObjectId, ref: "User", required: true },
    homeroomTeacher: { type: String, required: true },
    capacity: { type: Number, default: 0, min: 0, required: true },
    session: { type: Number, default: 0, min: 0, required: true },
    enrolledCount: { type: Number, default: 0, min: 0, required: true },
    members: [
      {
        userId: { type: Schema.Types.ObjectId, ref: "User", required: true },
        role: { type: String, enum: ["student", "teacher"], required: true },
        enrolledAt: { type: Date, default: Date.now },
        deletedAt: { type: Date, default: null },
        deletedBy: { type: String, default: null },
      }
    ],
  },
  { timestamps: true }
);

export const Course = mongoose.model<ICourse>("Course", courseSchema);

File: mirai_mobile/lib/models/course_model.dart
class CourseModel {
  final String id;
  final String name;
  final String? description;
  final String status;
  final String? startDate;
  final String? endDate;
  final String? createdBy;
  final String? homeroomTeacherId;
  final String? homeroomTeacher;
  final int session;
  final int capacity;
  final int enrolledCount;
  final String? createdAt;
  final String? updatedAt;

  final List<EnrolledStudent>? students;

  CourseModel({
    required this.id,
    required this.name,
    this.description,
    required this.status,
    this.startDate,
    this.endDate,
    this.createdBy,
    this.homeroomTeacherId,
    this.homeroomTeacher,
    this.session = 0,
    this.capacity = 0,
    this.enrolledCount = 0,
    this.createdAt,
    this.updatedAt,
    this.students,
  });

  factory CourseModel.fromJson(Map<String, dynamic> json) {
    return CourseModel(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'],
      status: json['status'] ?? 'not_yet',
      startDate: json['startDate'],
      endDate: json['endDate'],
      createdBy: json['createdBy'],
      homeroomTeacherId: json['homeroomTeacherId'],
      homeroomTeacher: json['homeroomTeacher'],
      session: json['session'] ?? 0,
      capacity: json['capacity'] ?? 0,
      enrolledCount: json['enrolledCount'] ?? 0,
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
      students: json['students'] != null
          ? (json['students'] as List).map((s) => EnrolledStudent.fromJson(s)).toList()
          : null,
    );
  }
}

class EnrolledStudent {
  final String id;
  final String name;
  final String email;
  final String? enrolledAt;

  EnrolledStudent({
    required this.id,
    required this.name,
    required this.email,
    this.enrolledAt,
  });

  factory EnrolledStudent.fromJson(Map<String, dynamic> json) {
    return EnrolledStudent(
      id: json['_id'] ?? json['id'] ?? json['userId'] ?? '',
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      enrolledAt: json['enrolledAt'],
    );
  }
}
